Fixes #11155 - Added support for aspect models

ShimShtein opened the pull request:
This is the first part of the work - it contains all changes that are needed in the existing environment.
Once it is merged, it will be possible to move on, and create a PR that will separate aspects from the existing host.

http://projects.theforeman.org/projects/katello/wiki/HostUnification

domcleal commented:
There's an existing PR, #2555 for this ticket - please close one and update Redmine as appropriate.

ShimShtein commented:
@domcleal, Set the link to this PR in Redmine, and closed the previous one. Am I missing anything else?

ohadlevy commented on app/models/host/managed.rb:
```diff
@@ -822,14 +854,26 @@ def available_template_kinds(provisioning = nil)
             end
 
     kinds.map do |kind|
-      ProvisioningTemplate.find_template({ :kind               => kind.name,
```
nitpick: not related change?

ShimShtein replied:
It's related. I want aspects be able to add filter options based on the information in the aspect.
I have added call to `#template_filter_from_aspects` which checks each of the aspects if it responds to `#template_filter_options`, and if it does - it can add another filtering.
You can see it implemented in the reference PR for the puppet environment: 
Base filter (without environment): https://github.com/theforeman/foreman/pull/2555/files#diff-2d53894341fd1cf84a823e3bd0855495R716
The addition (filter by puppet environment): https://github.com/theforeman/foreman/pull/2555/files#diff-21d16583cdd439b78dfb8ee84c3fc406R25

ohadlevy commented on app/services/host_aspects.rb:
```diff
@@ -0,0 +1,93 @@
+module HostAspects
+  class Configuration
```
should we break it down to each class in its own file?

ShimShtein replied:
It's kind of monolithic. If there was an option to create private classes - I would probably do it.
The main class here is the `Configuration` class, the 'Entry` is internal only, and the module just adds a singleton instance of the configuration.

ohadlevy commented on test/fixtures/host_aspects.yml:
```diff
@@ -0,0 +1 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/Fixtures.html
```
?

dLobatog commented:
@ShimShtein A link to the discussion & description of your plans on this area would help reviewers and anyone that stumbles upon this PR without context :smile:

ShimShtein commented:
@eLobato, Sure, here is our wiki page, we will try to maintain it as up to date as possible.
The page will contain all the relevant links.

http://projects.theforeman.org/projects/katello/wiki/HostUnification

jlsherrill commented on app/models/host/managed.rb:
```diff
@@ -673,8 +689,13 @@ def root_pass
+
     # do not copy system specific attributes
-    host = self.deep_clone(:include => [:host_config_groups, :host_classes, :host_parameters],
+    host = self.deep_clone(:include => [:host_config_groups, :host_classes, :host_parameters, :host_aspects] + aspects,
```
You may not want to clone all aspects.  Might make it opt out on a per aspect basis.

jlsherrill replied:
or a particular aspect may want certain attributes cloned but not others.  (Unsure if deep_clonable provides a way to override this on a per-class basis).

ShimShtein replied:
Theoretically it is implemented using the `#after_clone` hook. You can always remove anything from the object after it was cloned.
I agree this is not the best thing, but I would prefer a slimmer interface on the aspect model.

jlsherrill replied:
That is a tad ugly, but WFM!

On Sun, Sep 6, 2015 at 9:19 AM, Shimon Shtein notifications@github.com
wrote:

> In app/models/host/managed.rb
> https://github.com/theforeman/foreman/pull/2665#discussion_r38821355:
> 
> > ```
> >  # do not copy system specific attributes
> > ```
> > -    host = self.deep_clone(:include => [:host_config_groups, :host_classes, :host_parameters],
> > -    host = self.deep_clone(:include => [:host_config_groups, :host_classes, :host_parameters, :host_aspects] + aspects,
> 
> Theoretically it is implemented using the #after_clone hook. You can
> always remove anything from the object after it was cloned.
> I agree this is not the best thing, but I would prefer a slimmer interface
> on the aspect model.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman/pull/2665/files#r38821355.

ShimShtein commented:
Rebased and removed unnecessary fixture.

ShimShtein commented:
Missed a rubocop issue.

unorthodoxgeek commented on app/helpers/hosts_helper.rb:
```diff
@@ -478,4 +478,28 @@ def inherited_by_default?(field, host)
     return true unless params[:host]
     !params[:host][field]
   end
+
```
I like the concept of using the aspects to render tabs, it's quite eloquent.

That said, please check performance - make sure that queries are done as little as possible, for instance, when you bring in an aspect, if you can do that in one query for all aspects, prefer that. I didn't dive too deeply into the code, but it seems like the load_aspects method might do multiple queries where it can have only one, for instance.

unorthodoxgeek commented on app/services/host_aspects.rb:
```diff
@@ -0,0 +1,93 @@
+    SUBJECTS = [:configuration]
+
+    SUBJECTS.each do |subject|
+      define_method "register_#{subject}".to_sym do |aspect_name, aspect_model = nil, &block|
```
no need for the to_sym. anywhere. you can define_method with a string.

unorthodoxgeek commented on app/services/host_aspects.rb:
```diff
@@ -0,0 +1,93 @@
+
+    SUBJECTS.each do |subject|
+      define_method "register_#{subject}".to_sym do |aspect_name, aspect_model = nil, &block|
+        self.send('register', subject, aspect_name, aspect_model, &block)
```
why send?
should be 

``` ruby
register(subject, aspect_name, aspect_model, &block)
```

(register is private, thus this would work, self.register won't, because Ruby :stuck_out_tongue_closed_eyes:

unorthodoxgeek commented on app/services/host_aspects.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  class Entry
```
put in different file?
single class per single file plz :smile:

ShimShtein replied:
You are a second person that comments about it, I will separate them, although I have answered Ohad that these classes are monolithic and there is no use in this separation.

unorthodoxgeek commented on app/views/hosts/_form.html.erb:
```diff
@@ -22,6 +23,10 @@
       <li><a href="#params" id='params-tab' data-url='<%= current_parameters_hosts_path %>' data-url2='<%= puppetclass_parameters_hosts_path %>' data-toggle="tab"><%= _('Parameters') %></a></li>
       <li><a href="#info" data-toggle="tab"><%= _('Additional Information') %></a></li>
+
+      <% @tabs.keys.each do |tab| %>
```
I'd expect params, info, etc to be aspects as well. this adds clutter instead of removing it...

ShimShtein replied:
In my vision, host's aspect is any subject in the lifetime of a host that is not inventory - such as config, content e.t.c. Params and info are parts of the basic inventory - they should not belong to any aspect.

unorthodoxgeek replied:
then add "default aspects", this idea is very good IMHO, and now that it only adds clutter it misses the point. I might want to remove parameters altogether for hosts which are launched from a ready AMI (thus, they don't have a CM at all), I might decide to leave networking to the cloud platform, etc. this feature can allow a lot of flexibility, I don't see why we should limit it to new stuff.

ShimShtein replied:
Current tabs/properties would be sorted out too of course. I have already shown this for puppet configuration in PR #2555. In the [google groups thread](https://groups.google.com/forum/#!searchin/foreman-dev/host$20unification/foreman-dev/esDULZDlOK8/rmR-Rm9L3kIJ) I have mentioned all aspects that I can see in the near future.
This PR is about adding core support for aspects only. Once it gets merged, I will start working on moving puppet to an aspect.

unorthodoxgeek commented on app/views/hosts/_form.html.erb:
```diff
@@ -131,6 +136,9 @@
       </div>
 
+      <% @tabs.values.each do |tab| %>
+        <%= tab %>
```
what is in the values? actual HTML?

I think the more suitable way of doing this is to render partials, with values being the partial's path. This will have the added benefit of forcing us to use partials without passing local variables which are code smell...

ShimShtein replied:
Basically yes, the actual HTML. You can see the example of this usage here: https://github.com/theforeman/foreman/pull/2665/files#diff-b7708a9e000a6f55ba5cfb8553392dddR502
There is a difference between the construction of additional tabs (that do not require passing of `f` parameter, and aspect tabs. I am not sure if there could be other initialization process involved, so I left it open. I think your suggestion is a good idea, once we can agree on a common interface for all the tabs.

unorthodoxgeek replied:
I think the only local worth passing is "f" (the form builder object), other than that, everything smells fishy to me, @host or whatever should be an instance variable, not a local one (that is, in the tabs of the hosts#edit page)

domcleal commented:
Setting to WoC until the first feedback is addressed.  There's also a lot of missing test coverage for the additions to core views, helpers and models which needs to be added.

ShimShtein commented:
@domcleal About the testing: I will add testing to helpers, since they can be properly isolated, but it is a bit problematic to test the model and view until there will be actual model that implements an aspect.

domcleal commented:
You'll need to add one then, if only for the tests, else leave anything that can't be exercised out of the patch.

ShimShtein commented:
Rebased, config separated, more tests were added.
Katello tests are failing with "Permissions not found" error.

domcleal commented:
This is as clear as mud to me, I'm afraid.  Can you check if I even vaguely understand what these objects are?
1. HostAspect:
   - a joining model between Host::Managed and an execution model
   - with a type such as :configuration, and a host can have up to one impl per type
   - probably created via nested attributes on a host
2. HostAspect#execution_model:
   - some sort of implementation provided by a plugin
   - is an AR record, not a service/non-AR class
   - generally a subclass of HostAspectBase
   - has various optional methods to hook into Host::Managed methods or events, e.g. to provide ENC data or after cloning
3. HostAspectBase
   - abstract class for execution classes
   - why not call this execution class base, if that's what it is?
   - if I understand correctly, both the join model and the aspect (execution?) itself are called "aspect"
4. HostAspects:
   - a registration/management service
   - only permits one aspect type to be registered per predefined list of SUBJECTS

So if I was developing a foreman_foo plugin, which wanted to add a :foo aspect to Host::Managed, I'd first get :foo added to SUBJECTS so it's a known possible extension to Host::Managed, then define a FooAspect class, inheriting from HostAspectBase, create my foo_aspects table via a migration, and then hosts would get created (via view extensions and nested attrs) with the additional FooAspect resource and the join model HostAspect?

ShimShtein commented:
I'm glad that you have asked. I'll try to clarify as much as possible.

First a bit of theory:
Besides inventory of available hosts, Foreman handles different **SUBJECTS** of host lifetime management, such as configuration, content, subscriptions e.t.c.
**Aspect** is an implementation (as in brings it to the real world) of a single subject. For example, _configuration_ subject can be handled either by ChefAspect or by PuppetAspect.
This leads to a model where the administrator decides for each host, whether and how to handle each subject for this host. For example an administrator can decide for host1 that he wants to manage its configuration by using puppet, so he has to configure properties like puppet environment and classes on PuppetAspect instance in order to do that.

Now for the design:
1. HostAspects
   - Indeed a registration/management service.
   - Responsible for keeping a list of valid implementations for each subject in a given foreman installation.
   - For now permits only one implementation for each subject. i.e. you cannot install both chef and puppet as configuration aspects on a single Foreman.
2. HostAspectBase
   - Abstract class for aspect's model. For example: the model that stores properties that are relevant to puppet for a given host (_PuppetAspect_) should inherit from this base class.
3. HostAspect
   - a joining model between Host::Managed and each aspect.
   - Host has_many of those, in order to keep track which aspects are enabled for each host
   - List of them is an abstract list of all enabled aspects for a given host.
   - Provides a set of services (common interface/contract) that should be implemented by the aspect model, in order to participate in host's activities such as ENC data definition, host cloning e.t.c.
4. HostAspect#execution_model
   - Reference to the aspect implementation model
   - The aspect generally is a subclass of HostAspectBase, as described above.

I'll correct the example inline, as it will be clearer this way:
So if I was developing a foreman_foo plugin, which wanted to add a :foo aspect to Host::Managed, I'd first **decide which SUBJECT it implements, and if the subject is not listed, I'll add it to the list of available subjects** ~~get :foo added to SUBJECTS~~ **I'll register :foo using HostAspects,** so it's a known possible extension to Host::Managed, then define a FooAspect class, inheriting from HostAspectBase, create my foo_aspects table via a migration, and then hosts would get created (via view extensions and nested attrs) with the additional FooAspect resource. ~~and the join model HostAspect?~~ **And it will be available to abstract aspect consumers (such as views) via a dedicated HostAspect record.**

dmitri-d commented on app/helpers/hosts_helper.rb:
```diff
@@ -3,6 +3,7 @@ module HostsHelper
   include BmcHelper
+  include HostAspectsHelper
 
   def provider_partial_exist?(compute_resource, partial)
```
Probably useful to have a test for this method: it implicitly relies on a certain directory structure, and will break should that change. Perhaps pick a couple of most visible compute resources and test those paths?

ShimShtein replied:
Good idea, but I don't think it would be related to this commit though...

dmitri-d commented on app/models/concerns/host_aspects/managed_host_extensions.rb:
```diff
@@ -0,0 +1,75 @@
+      has_many :host_aspects, :foreign_key => :host_id, :inverse_of => :host do
+        def [](subject)
+          where(:aspect_subject => subject).first ||
+          select { |aspect| aspect.aspect_subject == subject}.first ||
```
Why would you need this in addition to line #12? Also, you can use #find on a collection to get the first element matching the predicate.

ShimShtein replied:
I need the second part of the `||` statement for the cases when the record is still new. In this case, the first part will return `[]` (as there are no records in the DB that match the condition) and all the actual records will be found in a local array that gets searched by the second part.
Probably it would be nicer to use `#find`, I will change that.

dmitri-d commented on app/models/concerns/host_aspects/managed_host_extensions.rb:
```diff
@@ -0,0 +1,75 @@
+      end
+      accepts_nested_attributes_for :host_aspects
+
+      def attributes
```
Have you tried saving an object with attributes extended in such a way? ActiveRecord may not be happy...

dmitri-d replied:
Nm, looks safe.

dmitri-d commented on app/models/host/managed.rb:
```diff
@@ -708,9 +717,14 @@ def root_pass
-    host = self.deep_clone(:include => [:config_groups, :host_config_groups, :host_classes, :host_parameters, :lookup_values],
-                           :except  => [:name, :mac, :ip, :uuid, :certname, :last_report])
+    host = self.deep_clone(:include => [:config_groups, :host_config_groups, :host_classes, :host_parameters, :lookup_values] + aspects,
+                           :except  => [:name, :mac, :ip, :uuid, :certname, :last_report ])
```
extra space?

witlessbird commented on app/models/host_aspect_base.rb:
```diff
@@ -0,0 +1,25 @@
+class HostAspectBase < ActiveRecord::Base
```
How is this class related to `HostAspect`, if at all?

dmitri-d replied:
Nm, read the explanation. Might make sense to rename `HostAspect`, as it's causing confusion.

dmitri-d replied:
There is no relation corresponding to `HostAspect.execution_model` here, is that intentionally?

dmitri-d replied:
Also, I can't seem to find the table definition for this?

ShimShtein replied:
> Might make sense to rename HostAspect, as it's causing confusion.

I am not too attached to this name, suggestions?

> There is no relation corresponding to HostAspect.execution_model here, is that intentionally?

It is not needed. `execution_model` is used to navigate to an actual record, when dealing with use cases that are working with an abstract definition of an aspect. The opposite direction is kind of pointless, since the user is already aware to the fact that he is dealing with an aspect.

> Also, I can't seem to find the table definition for this?

Tables for inherited classes (actual aspects) will be defined by the extensions that are adding some aspects to host. For example puppet extension will add a `PuppetAspect` model that will inherit from this class, and add a proper migration file.

domcleal replied:
> > Might make sense to rename HostAspect, as it's causing confusion.
> 
> I am not too attached to this name, suggestions?

Agreed, it's too ambiguous.  Perhaps the base class should be a plain "Aspect", which leaves the association model being a HostAspect, which makes a little more sense.

dmitri-d replied:
That would work. Also: I'm partial to not naming this abstraction an "aspect" as it can be confusing (traditionally aspect is an abstraction from aspect-oriented programming, which is rather common, if not in ruby world). The name of this pattern from the "gang of four" book is facet, so we could use that.

dmitri-d commented on app/services/host_aspects/configuration.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+
+    def registered_aspects
+      @aspects_registry.try(:dup) || {}
```
Nitpick: this is a bit too clever (took me a moment to realise what was going on here). Maybe have an accessor `aspects_registry` that always returns an initialized instance var?

Also, that way you'll have initialization code in one place, without any duplication.

dmitri-d commented:
What do you think would be a "normal" number of aspects in production? Have you tried measuring performance in calls that rely on aspects?

ShimShtein commented:
> What do you think would be a "normal" number of aspects in production? 

There would be at most one aspect per subject, so for now, I think it would be less than 10 aspects per host

> Have you tried measuring performance in calls that rely on aspects?

No I didn't, but since it's a simple 1:1 relation, these calls should be handled by the DB pretty effectively. I don't want to begin with complicating the code with optimizations, before we can identify a real bottleneck.

ShimShtein committed:
Fixes #11155 - Added support for aspect models

ShimShtein commented:
Rebased, and fixed @witlessbird's comments

domcleal commented:
1. How will an API user create a host with aspects?  If the UI uses nested attributes, that's OK, but they're an unfriendly format for the API, since it's designed for forms.  Also, how will API responses contain aspect attributes?
2. How will aspect attributes be stored on, and applied from host groups?
3. The mechanism of core code calling a method on every aspect, and it being passed to the execution model concerns me for a few reasons:
   1. Heavy use of `#respond_to?` rather than aspects registering extension points that they're interested in.  This would help to pick up on errors, internal API changes, allow deprecation warnings etc.
   2. No explicit ordering - plugins can't change the iteration order compared to another plugin, it's going to be fairly non-deterministic.
   3. For a plugin to add its own extension point, they'd have to extend the HostAspect model to pass through method calls?
   4. This constitutes a set of new, specific plugin APIs that will need documenting - including input/outputs of methods, since they're all different.

jlsherrill commented on app/services/host_aspects/configuration.rb:
```diff
@@ -0,0 +1,51 @@
+      aspect_subjects[subject] = entry.name
+    end
+  end
+end
```
I still question the need for 'Configuration'.  It would make the initial work more simple if there were not 'types' of aspects and would probably make the registration api more straight forward?

domcleal replied:
Yeah, I meant to raise this too in my last comment.  I think if I was adding an aspect from a plugin, I'd just ignore the categorisation.

jlsherrill commented:
> How will an API user create a host with aspects? If the UI uses nested attributes, that's OK, but they're an unfriendly format for the API, since it's designed for forms. Also, how will API responses contain aspect attributes?

@domcleal we had discussed adding support for this as a followup PR, would you prefer it in this PR?

domcleal commented:
> > How will an API user create a host with aspects? If the UI uses nested attributes, that's OK, but they're an unfriendly format for the API, since it's designed for forms. Also, how will API responses contain aspect attributes?
> 
> @domcleal we had discussed adding support for this as a followup PR, would you prefer it in this PR?

Yes, I would.  I don't really wish to merge a UI-only feature, and I think it needs some thought about how it would work.

If there are other things that are planned, please do file and associate Redmine issues so I know about it.

ShimShtein commented:
> How will an API user create a host with aspects? 

Let's look at this problem from user stories point of view:
_"I want to create a host with puppet aspect turned on and properties a, b and c set on it"_ - In this case we need to provide a unified way for the user to actually create this objects set. In order to do it the user has to specify a full JSON with all those properties, something like that:
`POST /hosts`

``` JSON
"host": {
  "name":"awesome.com",
  "puppet": {
    "environment":"prod",
    "proxy_id": 1
   },
   "other_aspect":{
    "prop1": "val1"
   }
}
```

The total number of properties will not be greater than what we have today, they are only grouped differently (which makes more sense, if you ask me).
_"I want to add/update puppet aspect for host X" - Aspects are plain models, so they will get their own API calls to treat them independently.
`PUT /hosts/X/puppet_aspect`

``` JSON
"puppet_aspect": {
    "environment":"prod",
    "proxy_id": 1
}
```

> If the UI uses nested attributes, that's OK, but they're an unfriendly format for the API, since it's designed for forms. 

About the "friendliness" - I am probably missing something, but the JSON I presented is pretty simple, I can't see any simplification in the API layer either...

> Also, how will API responses contain aspect attributes?

Exactly the same way as we expect the user to input them - Like I showed in the `POST /hosts` example. That means including them in the host by default (maybe with some kind of exclude option to opt out).

> How will aspect attributes be stored on, and applied from host groups?

After the full redesign, the same way they are managed in hosts - hostgroups will have aspects attached to them. The same concept will go on for each abstract (core) object we have in foreman.

> Heavy use of #respond_to? rather than aspects registering extension points that they're interested in.

In compiled language I would create an interface that the aspect will inherit, here I am using a naming convention for that. I can get rid of `#respond_to?`  by adding empty method stubs to HostAspectBase (or in its new name Aspect).

> No explicit ordering - plugins can't change the iteration order compared to another plugin, it's going to be fairly non-deterministic.

There should be no ordering between the plugins - they all just add more info to host. I didn't see a case where the order will matter.

> For a plugin to add its own extension point, they'd have to extend the HostAspect model to pass through method calls?

Sorry, didn't understand the use case, can you elaborate more?

> This constitutes a set of new, specific plugin APIs that will need documenting - including input/outputs of methods, since they're all different.

Agreed.

>  I think if I was adding an aspect from a plugin, I'd just ignore the categorisation.

The categorization serves two purposes. From the user point of view, it protects from ambiguity, for example the configuration will be managed either by puppet or by chef but not both. From architecture point of view, we will be able to extract common properties later on, and put them outside the implementing aspect. So if both puppet and chef share the concept of environment, it could be put away into something common.

ShimShtein committed:
Refs #11155 - Set of changes that I found after creating puppet aspect

ShimShtein commented:
Main changes:
- Removed HostAspect model and made the list in-memory
- Added API support.
- Created an extension point via DSL for adding properties to cloning process
- Removed `#responds_to?` pattern and moved it to defaults in base class.
- Removed aspect's category field

dLobatog commented on test/unit/host_test.rb:
```diff
@@ -2186,6 +2187,7 @@ def to_managed!
   end
 
   test '#primary_interface works during deletion' do
+    skip 'Waiting for rails4 '
```
Why?

ShimShtein replied:
This test actually tests that ActiveRecord's caching on relations keeps instances of all interfaces, even if they are deleted. 
I had to add `after_save :clear_association_cache` to our host model in order to fix references issue after save ( `host.some_aspect.host.object_id` is not equal to `host.object_id`). 
In Rails 4 the issue with the relations is fixed, so then it will be logical to remove the `after_save` hook and put the test back.

dLobatog commented on test/unit/host_test.rb:
```diff
@@ -1954,6 +1954,7 @@ def to_managed!
   end
 
   test "#info ENC YAML omits root_pw when password_hash is set to Base64" do
+    skip 'Now after save, the password is ecrypted, so the host.info never shows the unencrypted password'
```
Why skip instead of removing the test directly?

ShimShtein replied:
I wasn't sure that the change in behavior is a valid one, so I wanted another pair of eyes on it before I remove the test and confirm the behavior change.
And again, as in my previous comment, in rails 4 the `after_save` hook should be removed. So I am not sure that the test will remain invalid then.

dmitri-d commented on app/services/host_aspects.rb:
```diff
@@ -0,0 +1,11 @@
+module HostAspects
```
If the plan is to use aspects not just with hosts, but also hostgroups (possibly something else?), it would be better not to use 'Host' in the name of the module.

dLobatog replied:
How about using the singular - `Aspect`? The module is not implementing multiple aspects as far as I can see.

ShimShtein replied:
This class represents configuration for all available aspects, so plural form is kind of logical here.

dmitri-d commented on app/services/host_aspects.rb:
```diff
@@ -0,0 +1,11 @@
+      @configuration ||= HostAspects::Configuration.new
+    end
+
+    def configure
```
where's this being used?

ShimShtein replied:
When you want to configure your aspect, it will be in the initializer code:

``` Ruby
HostAspects.configure do |config| 
  config.register 'MyAspect'
end
```

dmitri-d replied:
Why use a block here?

ShimShtein replied:
You are right, it should be:

``` Ruby
HostAspects.configure do
  register 'MyAspect'
end
```

dmitri-d commented on app/helpers/host_aspects_helper.rb:
```diff
@@ -0,0 +1,29 @@
+  def host_additional_tabs(host)
+    base_tabs = {}
+    aspect_tabs = load_aspects(host)
+    base_tabs.merge!(aspect_tabs)
```
if you are trying to minimize the local state (like here: https://github.com/theforeman/foreman/pull/2665/files#diff-13b828e2868627bafc7af5e036628234R7), why not: `base_tabs.merge!(load_aspects(host))` :P

dmitri-d commented on app/helpers/host_aspects_helper.rb:
```diff
@@ -0,0 +1,29 @@
+module HostAspectsHelper
+  # override host's tabs to add all aspects as tabs
+  def host_additional_tabs(host)
```
this is used not just in UI views, but also to render responses to API calls, where 'tabs' sounds confusing.

ShimShtein replied:
Agreed, is `host_additional_views` better?
It returns hash of ids to template names.

dmitri-d commented on app/views/api/v2/hosts/show.json.rabl:
```diff
@@ -25,3 +25,16 @@ end
   extends "api/v2/config_groups/main"
 end
+
+host_additional_tabs(@host).each do |id, tab|
```
Should the index view also show the information from aspects (I know it's pretty minimal atm and doesn't have all of the data the 'show' view)?

ShimShtein replied:
Right, little bit untrivial, but doable. I'll add it too.

dLobatog commented on app/models/host_aspect_base.rb:
```diff
@@ -0,0 +1,36 @@
+    end
+  end
+
+  def info
```
Comments on how to override these generic methods would be useful. Maybe something should be raised instead of returning a default, or are these ones meant to be used?

ShimShtein replied:
I'll add comments with explanation what each method is used for, sure.
They should not throw exceptions, because you don't have to override them if you don't have what to add to the base method. The method will be called for each aspect, so if one of them throws an exception - the whole method will fail.

dLobatog commented on test/aspect_test_helper.rb:
```diff
@@ -0,0 +1,8 @@
+Spork.prefork do
+  unless ActiveRecord::Base.connection.table_exists?(:test_aspects)
+    ActiveRecord::Base.connection.create_table :test_aspects do |t|
```
Is it necessary to have a table to register an aspect (that's what the classes that use this helper do, right?)? How about mocking all db calls if all the table contains is the `host_id`?

ShimShtein replied:
When using postgres, the models get validated against a db. You can try to mock it out, but it will be just too hard, you will probably need to mock the whole undeluying database adapter instance including all its methods.

dLobatog commented on test/unit/host_aspect_configuration_test.rb:
```diff
@@ -0,0 +1,47 @@
+  test 'gives readonly access to the registry' do
+    config = HostAspects::Configuration.new
+    config.register 'MyCoolAspect', :my_cool_model
+    hash = config.registered_aspects
```
I'm not sure I follow why saving `registered_aspects` in a variable then modifying it and checking the original `registered_aspects` tests `registered_aspects` is read only? That'd be a valid test in a pass-by-reference language but Ruby is pass by value. Would this work if this is what you're trying to test?? 

``` ruby
assert_difference('config.registered_aspects.count', 0) do
  config.registered_aspects.delete(:MyCoolAspect)
end
```

ShimShtein replied:
Yep, this is a better implementation for the test.

dLobatog commented on test/unit/host_aspect_configuration_test.rb:
```diff
@@ -0,0 +1,47 @@
+    test 'defaults initialization' do
+      config = HostAspects::Configuration.new
+      config.register 'MyCoolAspect'
+      res = config.registered_aspects[:MyCoolAspect]
```
What's `res`? Is there a better name?

dLobatog commented on test/unit/host_aspect_configuration_test.rb:
```diff
@@ -0,0 +1,47 @@
+require 'test_helper'
+
+class HostAspectConfigurationTest < ActiveSupport::TestCase
+  class MyCoolClass
```
Bikeshedding here, but I'd `s/MyCool/AspectTest/g` `s/my_cool/aspect_test` this file to make the intent clear

dLobatog commented on test/unit/orchestration/dhcp_test.rb:
```diff
@@ -65,7 +65,8 @@ def teardown
 
   test "provision interface DHCP records should contain filename/next-server attributes" do
-    ProxyAPI::TFTP.any_instance.expects(:bootServer).returns('192.168.1.1')
+    ProxyAPI::TFTP.any_instance.unstub(:bootServer)
```
At this point, when is `bootServer` stubbed? Also why not leaving the `expects` so that it tests `bootServer` is called?

ShimShtein replied:
it is stubbed in `#setup`, and for some reason you can't override the original stubbed value, unless you unstub it first.
I am creating the same expectation in the next line. The problem is that it is called more than once, so I needed to change it anyway. I am not sure that `expects.at_least_once` is a necessary check in this test.

dLobatog commented on app/models/host_aspect_base.rb:
```diff
@@ -0,0 +1,36 @@
+    aspect_attributes
+  end
+
+  def self.populate_fields_from_facts(host, importer, type, proxy_id)
```
This method particularly worries me a bit - is the `proxy_id` necessary? Can it be fetched from the importer or the host instead? It makes a ton of method signatures and calls change.

ShimShtein replied:
It is actually broken in our code: in `Host::Base` the definition of `self.import_host_and_facts` doesn't need a proxy, but in `Host::Managed` it already does. That means that the inheritance doesn't work already (you have to treat Base and Managed differently anyway).
I would like to see this mechanism refactored, but I didn't want to do it inside this PR, for the sake of less complication.

dLobatog replied:
The definition on `import_host_and_facts` on `Host::Managed` doesn't _need_ a proxy, it's nil if it can't be found. Otherwise you couldn't submit facts from non-registered proxies, or from [foreman_ansible](https://github.com/dlobatog/foreman_ansible)

ShimShtein replied:
That's right, but it won't work the other way around - if you try to call the method while specifying proxy on something other than Host::Managed.
Maybe the better option will be to create some kind of context object that will be populated by the host, and then passed to this method and the whole chain that it is calling. This context object can be a simple options hash.

lzap commented on app/services/host_aspects/configuration.rb:
```diff
@@ -0,0 +1,29 @@
+      @aspects_registry ||= {}
+    end
+  end
+end
```
Nitpick: Can you configure your editor to preserve EOFEOL?

lzap commented:
Should Copute Resource association be also extracted as an aspect? I don't see that one in the design document.

ShimShtein commented:
> Should Copute Resource association be also extracted as an aspect? I don't see that one in the design document.

@lzap, It can be... Probably I am missing something, but I can't see why it is different from puppet aspect.
My guideline is to leave the bare host to deal only with inventory.

ShimShtein committed:
Refs #11155 - Last changes after rebase

jlsherrill commented on app/models/host/managed.rb:
```diff
@@ -640,6 +673,20 @@ def apply_inherited_attributes(attributes, initialized = true)
+    end
+
+    attributes
+  end
```
Is this method used anywhere?  or planned to be used?

ShimShtein replied:
Nice catch! Leftover from previous implementation.

jlsherrill commented on app/models/host/managed.rb:
```diff
@@ -510,13 +532,18 @@ def attributes_to_import_from_facts
+
     operatingsystem.architectures << architecture if operatingsystem && architecture && !operatingsystem.architectures.include?(architecture)
     self.save(:validate => false)
   end
```
Why does this method actual need a proxy_id  (and its base.rb counterpart)?  You aren't actually using that here, except to pass to the aspect's populate_fields_from_facts method, will it actually be used there?  (Since its not used today I'm not sure why it would be needed in the aspect)

ShimShtein replied:
In puppet implementation proxy_id is needed, as I wrote to @dLobatog, it should be refactored regardless this PR.

jlsherrill replied:
yeah, i was more curious about this particular method, but i see now https://github.com/ShimShtein/foreman/blob/puppet-aspect/app/models/puppet_aspect.rb#L247  uses it in your puppet implementation.  I guess that was moved from Host::Managed.import_host_and_facts  to populate_fields_from_facts within the puppet aspect and that is the cause.  :+1:

jlsherrill commented on app/views/api/v2/hosts/show.json.rabl:
```diff
@@ -25,3 +25,16 @@ end
+  else
+    child tab => id.to_sym do |aspect|
+      class_name = tab.class.name.underscore
+      extends "api/v2/#{class_name.pluralize}/base"
```
hrmmm, all the current katello views are under  katello/api/v2/class_names/ (to follow rails module conventions), but this will look for them under api/v2/katello/class_names/.  Maybe we should provide a way to override this per aspect? 

 lets chat about this some more.

jlsherrill replied:
oh but  actually app/views//katello is a 'view path', so really in our case we likely want just the class base name (i.e.  MyAspect rather than  MyModule::MyAspect)

ShimShtein commented:
@jlsherrill I was trying to mimic the paths from the UI views to reduce the amount of declarations for each aspect (facet). I am actually using the same declaration in the UI and API, so both of them should have the same file structure.

jlsherrill commented:
@ShimShtein right, but to achieve this structure and follow rails conventions I'd have to put my views in  `/app/views/MODULE/api/v2/MODULE/CLASS/` specifically for this case, when none of my other views are like that.

ShimShtein committed:
Refs #11155 - Rename aspect to facet

ShimShtein commented:
@jlsherrill Added `#demodulize` to the rabl views, now it will be: `/app/views/MODULE/api/v2/CLASS/` which is consistent with your current file scheme.

ShimShtein commented:
@witlessbird @dLobatog Did the renaming, and fixed all other issues that were raised.

jlsherrill commented:
@ShimShtein tested this by integrating it into the katello content aspect, and found a couple issues related to namespacing/engines:

a) https://github.com/theforeman/foreman/pull/2665/files#diff-54b013ef0e142d15196526e364ad1a1bR8
is called at class loading time, but that is too late for an engine to be loaded.   Changing this to occur at registration time:  https://github.com/theforeman/foreman/pull/2665/files#diff-b5966ce9893730ab9aec9dcd40752e8dR7  (and modifying to only do it for that one facet worked great).

b)  A few of the references in managed_host_extensions.rb don't work when the facet name != the full class name

I put changes for a & b here:  https://gist.github.com/jlsherrill/b5c04c45935169e66201

c) In order to make it all work with the way the code is post patch, i had to do:

```
      Facets.configuration.register(:content_aspect, 'katello/host/content_aspect') do
        extend_model 'katello/concerns/content_aspect_host_extensions'
      end
```

Would it make more sense for the user to actually pass in the class?  i.e.:

```
      Facets.configuration.register(:content_aspect, ::Katello::Host::ContentAspect') do
        extend_model ::Katello::Concerns::ContentAspectHostExtensions
      end
```

Thats more what i would expect from a plugin API.

ShimShtein commented:
@jlsherrill Fixed all issues, and added the option to use actual classes.

ShimShtein committed:
Refs #11155 - Katello remarks

jlsherrill commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,50 @@
+
+        # include all facet attributes by default
+        host_facets_with_definitions.each do |facet, facet_definition|
+          hash["#{facet_definition.name}_attributes"] = facet.attributes.reject { |key, _| %w(created_at updated_at).include? key }
```
the  my_facet_attributes method doesn't seem like it exists for me, added:

```
        define_method("#{facet_config.name}_attributes") do
          self.send(facet_config.name).attributes
        end
```

in the included register_facet_relation method and that seemed to have fixed it

ShimShtein replied:
Actually I did it in my [puppet implementation](https://github.com/theforeman/foreman/pull/2555/files#diff-1fcf71199e401fa452c2a82f900b6ed6R12), but I don't remember why...

I'll add it to `#register...` method.

jlsherrill commented on app/helpers/facets_helper.rb:
```diff
@@ -0,0 +1,27 @@
+      if facet_definition.tabs.is_a? Hash
+        tab_definitions.merge!(facet_definition.tabs)
+      else
+        method_result = send(facet_definition.tabs, host)
```
This seems to indicate that if i pass  :my_facet_tabs as a tab, that this will call  `send(:my_facet_tabs)` implying that that method is defined in my helper that i have also specified (:my_facet_helper), but i don't see where that helper would be 'included'.  When i try to use this way of doing tabs i get undefined method because the 'my_facet_tabs' method doesn't exist

ShimShtein replied:
All helpers should be included by default in Rails. Maybe there is a problem with the name of the helper?

jlsherrill commented on app/views/api/v2/hosts/show.json.rabl:
```diff
@@ -25,3 +25,16 @@ end
+      extends "api/v2/#{class_name.pluralize}/base"
+    end
+  end
+end
```
Any thoughts on adding support for base.json.rabl as well so we can modify the index results?

jlsherrill replied:
errr i mean main.json.rabl

jlsherrill commented on app/views/api/v2/hosts/show.json.rabl:
```diff
@@ -25,3 +25,16 @@ end
+host_additional_views(@host).each do |id, tab|
+  if tab.is_a? String
+    node do |host|
+      partial "api/v2/#{tab}", :object => host
```
This doesn't actually work with namespaced views, which you generally do in an engine.  we'd need 'katello/api/v2/#{tab}'

I dont' know that i'm a huge fan of re-using the UI 'tab' mechanism for api requests as well.  I don't think a 'tab' in the ui would necessarily correspond to a rabl in the json.  Also for the katello case at first we don't even want to add any UI tabs but do need the parts exposed in the API.

ShimShtein replied:
I see that the "reuse the same structure in API and UI" approach doesn't play well... I think I'll just create two extra extension points to the configuration: `#api_host_view` and `#api_host_list`

jlsherrill commented:
Also, any thoughts on being able to extend the apipie documentation for hosts?

dLobatog commented on app/helpers/facets_helper.rb:
```diff
@@ -0,0 +1,27 @@
+    base_tabs
+  end
+
+  def load_facets(host)
```
This is not being called from outside `host_additional_views`, might be better if it's private

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,50 @@
+      host_facets_with_definitions.keys
+    end
+
+    def host_facets_with_definitions
```
Rename to `facet_with_definitions`, as this is called on Host instances so right now it's `host.host_facets_with_definitions`

dLobatog replied:
Comment how the output of this method looks like:

``` ruby
{ 
  host.puppet_facet => Facet::Entry # Puppet facet entry 
}
```

dLobatog commented on app/models/concerns/facets/managed_host_extensions.rb:
```diff
@@ -0,0 +1,50 @@
+      end
+    end
+
+    def host_facets
```
Same as `facets_with_definitions`

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -1,5 +1,19 @@
+  cattr_accessor :cloned_parameters
+  self.cloned_parameters = {}
+
+  def self.include_in_clone(*attributes)
```
@ShimShtein How about extracting include/exclude to a module that can be also used in Hostgroup? That would be a first, easy to merge PR. Also instead of initializing cloned_parameters to an empty hash, `Hash.new([])` sets the default so that these two methods could just be `cloned_parameters[:include] += attributes`

dLobatog commented on app/models/host_facet_base.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  # Add or change search criteria for finding a configuration template.
+  def provisioning_template_options(base_options)
```
We can remove this `base_options` if options are not being to changed, and just remove a {}

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -336,6 +350,10 @@ def provisioning_template(opts = {})
     opts[:environment_id]     ||= environment_id
 
+    host_facets.each do |facet|
+      opts = facet.provisioning_template_options(opts)
```
If `provisioning_template_options` returns a Hash, then you would call `opts.merge!(facet.provisioning_template_options)` here

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -427,6 +445,10 @@ def info
     info_hash['parameters'] = param
     info_hash['environment'] = param["foreman_env"] if Setting["enc_environment"] && param["foreman_env"]
 
+    host_facets.each do |facet|
```
Maybe this whole `info` method should be part of the Puppet facet, and we can extract it in that PR instead of offering an endpoint in the generic host facet

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -736,6 +771,11 @@ def clone
     end
     host.refresh_global_status
+
+    host.host_facets.each do |facet|
```
`puppet_status = 0` was moved to `refresh_global_status` so we don't _need_ the facet to support this at the moment, let's wait until it's needed to add it

dLobatog commented on app/models/host_facet_base.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  # This method will be called after a host with all its facets is cloned.
+  def after_clone
```
If it's not used up there let's remove the endpoint in the generic facet as well

dLobatog commented on app/models/host/managed.rb:
```diff
@@ -790,6 +830,11 @@ def smart_proxy_ids
       ids << p
     end
+
+    host_facets.each do |facet|
```
:+1:

dLobatog commented on app/models/host_facet_base.rb:
```diff
@@ -0,0 +1,52 @@
+class HostFacetBase < ActiveRecord::Base
+  self.abstract_class = true
+
+  def self.inherited(subclass)
```
Shouldn't `belongs_to_host` without the wrapping method work? It's the way it works on Taxonomy

ShimShtein commented on app/services/facets.rb:
```diff
@@ -0,0 +1,11 @@
+module Facets
```
``` Ruby
Rails.application.config.to_prepare do
  Facets.configuration.register(:puppet_facet) do
    extend_model :puppet_host_extensions
    add_helper :puppet_facet_helper
    add_tabs :puppet_tabs
    api_view nil, 'api/v2/puppet_facets/single_host_view'
    api_pie :puppet_properties, PuppetFacetsController
  end
end
```

dLobatog replied:
Can we pass classes instead for most of these? There is no need to pass the symbols for most:

``` ruby
Facets.configuration.register(PuppetFacet) do 
  extend_model PuppetHostExtensions
  add_helper PuppetFacetHelper
  add_tabs :puppet_tabs
  api_view { :list => 'api/v2/puppet_facet/hosts', :single => 'api/v2/puppet_facet/hosts'}
  api_docs :puppet_properties, PuppetFacetController, 'a description goes here'
end
```

.register could take a class, and a symbol only for those cases where the class name would be too long or undesirable to convert into a method. For instance, Katello::Subscriptions could be called as `.register(Katello::SubscriptionFacet, :subscription_facet)` rather than `.katello_subscription_facet`?

dLobatog commented on app/services/facets/configuration.rb:
```diff
@@ -0,0 +1,31 @@
+
+    private
+
+    attr_writer :facets_registry
```
This is creating `def facets_registry=` privately but it's not used within this class, I think it can be removed

dLobatog commented on app/services/facets/configuration.rb:
```diff
@@ -0,0 +1,31 @@
+module Facets
+  class Configuration
```
This logic can go in `Facets`, and then `facets_registry` can be called just `registry` as it would be called from other classes as `Facets.registry`

dLobatog commented on app/services/facets/configuration.rb:
```diff
@@ -0,0 +1,31 @@
+      Facets::ManagedHostExtensions.register_facet_relation(Host::Managed, entry)
+    end
+
+    def definition_for(facet_instance)
```
This is not called from anywhere, might not be needed?

dLobatog commented on app/services/facets/entry.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    def add_helper(facet_helper)
+      self.helper = facet_helper.try(:to_sym)
```
`try :to_sym` would go away if facet_helper is a class

dLobatog commented on app/services/facets/entry.rb:
```diff
@@ -0,0 +1,53 @@
+      end
+    end
+
+    def initialize(facet_name, facet_model)
```
Invert the order of arguments here and find the facet_name from the model, it would fit most cases

ShimShtein commented on app/services/tax_host.rb:
```diff
@@ -148,7 +148,15 @@ def non_inherited_ids(v1 = self.selected_ids, v2 = self.inherited_ids)
-      hosts.map(&key).uniq.compact
+      ids = []
+      hosts.each do |host|
+        host.host_facets.each do |facet|
```
@dLobatog , I will create a compatibility properties declaration in `Facets::Entry`, that will add pass through with deprecation properties in host object. This way we will not need to change tax_host right now.

``` Ruby
#on Host::Managed
def environment_id
   deprecation_warning
   puppet_facet.environment_id if puppet_facet
end
```

ohadlevy commented:
@ShimShtein can you rebase please?

ShimShtein commented:
@ohadlevy, Me and @dLobatog have decided that it would be easier to push this code in parts. First part is #2944  (Facet registration). Once it gets merged, I have another branch which will add model extension functionality. After those two, I will complete all of the functionality by pushing third PR with UI/API extensions
